artiq/dashboard/moninj/device_manager.py

```diff
@@ -0,0 +1,319 @@
+from artiq.dashboard.moninj.util import setup_from_ddb
+from artiq.dashboard.moninj.widgets.dac import DACWidget
+from artiq.dashboard.moninj.widgets.dds import DDSWidget
+from artiq.dashboard.moninj.widgets.ttl import TTLWidget
+from artiq.tools import logger
+
+if typing.TYPE_CHECKING:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+class DeviceManagerProxy(ICommMonInj):
+    def __init__(self, notify_port: int):
+        self.notify = Notifier(dict({"monitor": dict(), "injection_status": dict(), "disconnect": []}))
+        self.ddb = dict()
+        self.notify_port = notify_port
+        self.core_connection = CommMonInj(self.on_monitor, self.on_injection_status,
+                                          self.on_disconnect)
+        self.ddb_notify = Subscriber(notifier_name="devices",
+                                     notify_cb=self.on_notify,
+                                     target_builder=self.on_notify_build)
+
+    async def on_notify(self, mod):
+        if mod["action"] != "init":
+            await self.reconnect()
+
+    async def on_notify_build(self, init):
+        self.ddb = init
+        await self.connect_coredev()
+        return self.ddb
+
+    def on_monitor(self, channel: int, probe: int, value: int):
+        if channel not in self.notify["monitor"].raw_view:
+            self.notify["monitor"][channel] = dict()
+        self.notify["monitor"][channel][probe] = value
+
+    def on_injection_status(self, channel: int, override: int, value: int):
+        if channel not in self.notify["injection_status"].raw_view:
+            self.notify["injection_status"][channel] = dict()
+        self.notify["injection_status"][channel][override] = value
+
+    def on_disconnect(self):
+        self.notify["disconnect"].append(None)
+
+    async def reconnect(self):
+        await self.stop()
+        await self.connect()
+
+    async def stop(self):
+        try:
+            await self.ddb_notify.close()
+            await self.core_connection.close()
+        except:
+            pass
+
+    async def connect(self):
+        await self.ddb_notify.connect("localhost", self.notify_port)
+
+    async def connect_coredev(self):
+        try:
+            await self.core_connection.connect(self.ddb["core"]["arguments"]["host"], 1383)
+        except asyncio.CancelledError:
+            raise
+        except:
+            logger.error("failed to connect to core device moninj", exc_info=True)
+            await asyncio.sleep(10.)
+
+    def monitor_probe(self, enable: int, channel: int, probe: int):
+        if "_writer" in self.core_connection.__dict__:
+            self.core_connection.monitor_probe(enable, channel, probe)
+
+    def monitor_injection(self, enable, channel, overrd):
+        if "_writer" in self.core_connection.__dict__:
+            self.core_connection.monitor_injection(enable, channel, overrd)
+
+    def inject(self, channel, override, value):
+        if "_writer" in self.core_connection.__dict__:
```

Comment:
Shouldn't be called core_connection since it doesn't connect to the core device anymore.